schema_utils.py

```diff
@@ -185,6 +204,203 @@ def get_validator(validator_id):
     return _Validators.get(validator_id)
 
 
+def is_numeric_operand(string):
+    """Checks if the given string represents a valid number.
+
+    Args:
+        string: str. The operand to be validated.
+
+    Returns:
+        bool. Whether the given string is a valid numeric operand.
+    """
+    try:
+        float(string)
+        return True
+    except ValueError:
+        return False
+
+
+# Given string must be a singular english alphabet or a valid greek letter.
+IS_ALGEBRAIC_OPERAND = lambda string: (
+    (string.isalpha() and len(string) == 1) or string in GREEK_LETTERS.keys())
+
+IS_OPERAND = lambda c: is_numeric_operand(c) or IS_ALGEBRAIC_OPERAND(c)
+
+
+def contains_balanced_brackets(expression):
+    """Checks if the given expression contains a balanced bracket sequence.
+
+    Args:
+        expression: str. A math expression (algebraic/numeric).
+
+    Returns:
+        bool. Whether the given expression contains a balanced
+            bracket sequence.
+    """
+    openers, closers = '({[', ')}]'
+    stack = []
+    for character in expression:
+        if character in openers:
+            stack.append(character)
+        elif character in closers:
+            if len(stack) == 0:
+                return False
+            top_element = stack.pop()
+            if openers.index(top_element) != closers.index(character):
+                return False
+    return len(stack) == 0
+
+
+def tokenize(expression):
+    """Tokenizes the given math expression into separate components.
+
+    Args:
+        expression: str. A math expression (algebraic/numeric).
+
+    Returns:
+        list. A list of tokens present in the expression.
+    """
+
+    # Removing whitespace.
+    expression = expression.replace(' ', '')
+
+    # Replacing greek letters with corresponding symbols.
+    for letter in GREEK_LETTERS:
```

Comment:
Seems risky; an expression may use both alpha and a, for example. (Add backend tests for this case.)

Reply:
Yes, but since this is just a structural check, it doesn't matter. Initially, I'd used 'α' instead of 'a' but that was causing some issues while validating the postfix expression since it would get converted to its unicode equivalent which is not a single character and would, therefore, be considered as an invalid token. With this approach, something like `alpha^a + beta^b` would be converted to `a^a + b^b` which is structurally equivalent and simpler to handle.
I'll still add a backend check just in case.

Reply:
> Ah yes, I see. I suggest that you add a note to that effect in the comment.

Added

> However, I think using expression.replace() is tricky, and that you should be doing the substitution as part of tokenization instead.

Sorry if I'm misunderstanding, but it _is_ a part of tokenization, isn't it?

> Otherwise, what if the expression contains something like "balpha"?

That would be converted to "ba" which is still structurally invalid. I don't think there are any examples wherein greek words cause false positives to arise.

> More generally, why don't you use the existing Backus-Naur grammar at the bottom of http://asciimath.org/ to construct the parse tree (something like this perhaps)? That way, you won't have any discrepancies between backend and frontend.

I think, to check for syntactic validity, there's no need to generate a parse tree, we can simply convert infix to postfix and evaluate the postfix -- which is what I've done atm.

Reply:
>> However, I think using expression.replace() is tricky, and that you should be doing the substitution as part of tokenization instead.
> Sorry if I'm misunderstanding, but it is a part of tokenization, isn't it?

Not really. Tokenization means splitting the string into its constituent parts. It doesn't involve replacing or overwriting parts of the original string -- if you tokenize b * alpha you should get ['b', '*', 'alpha']. Substitution only comes after tokenization but is usually not something that's done.

> That would be converted to "ba" which is still structurally invalid. I don't think there are any examples wherein greek words cause false positives to arise.

Is it structurally invalid? I typed "ba" into the demo at http://asciimath.org/ and it works. There is a legitimate ambiguity about whether to interpret "balpha" as "b * alpha" or "b * a * l * p * h * a" so you need to make sure that your results here always match whatever asciimath does, and there needs to be proof of that. 

> I think, to check for syntactic validity, there's no need to generate a parse tree, we can simply convert infix to postfix and evaluate the postfix -- which is what I've done atm.

The onus is on you to prove formally/rigorously that this works in all cases, not on me (or your users, if you push this to production) to find counterexamples and run into errors/frustrations. The proof will be easier if you validate against the already-established grammar using standard methods for handling Backus-Naur or context-free grammar parsing, since by definition this will work. But if you have a parsing approach that is completely different, then it needs a formal proof of equivalence.

More specifically, there needs to be a formal grammar for what "algebraic expression" is and you need to parse according to that grammar. If that grammar is the same as the asciimath grammar, then you can default to that. If it's a different grammar you need to specify it here, make clear that it's a proper subset of asciimath, and parse according to that second grammar.

Note that the grammar/proof should be included in the relevant docstring. This is important for future maintainability, since it allows us to update the grammar and/or update the proof if the definition of validity changes in the future.

Let me know if you have any questions about this.

Reply:
As per discussion here, the approach of validation has been updated to generating a parse tree instead of simply converting from infix to postfix.

About the concatenation of operands -- With the new implementation, concatenation of operators is not allowed, i.e., `ab` != `a * b`. This would imply that there is no ambiguity between variables, greek letters, and math functions. So, "balpha" would be treated as `b * alpha` and not `b * a * l * p * h * a`.